### ISSUE TITLE
Function to Execute Raw Bytecode

### ISSUE THREAD

**Issue by cwinsloweaton** [ethereum/py-evm#306]
Was asked to open an issue here, hope this is alright.

It would be pretty useful to have a function that takes bytecode and data and executes the bytecode along with the data. Ethereumjs-vm has a function [runCode()](https://github.com/ethereumjs/ethereumjs-vm#vmruncodeopts-cb) that does this and works pretty well, but I'd rather be using Python.

**Comment by pipermerriam** [comment]
This strikes me as a nice convenience API to have available.  I think this fits well as a new API on the `VM` class or as a stand-alone that takes a VM class as one of it's arguments.  There are a few additional parameters that would need to be filled in, though I suspect we can supply sane defaults for most of them (`sender/coinbase/anything-else-on-the-message-object`).

**Comment by pipermerriam** [comment]
```python
def execute_raw_bytecode(vm, bytecode, data=b'', ...):
    ...
```

Extra arguments would be something like every property of the `evm.vm.message.Message` object, with all of them defaulted to empty values or whatever other defaults we can come up with.

**Comment by lrettig** [comment]
This is up my alley, given the work I've been doing on eWASM. I'll take a crack at it.

**Comment by lrettig** [comment]
How's this for a set of defaults?

These mostly come from https://github.com/ethereumjs/ethereumjs-vm/blob/master/lib/runCode.js (except that it does not specify a default `gasPrice`):

- gas: 0xffffff
- gas_price: 0x01
- to: 0x0
- sender: 0x0
- value: 0x0
- data: None
- code: None

These defaults are already specified in `evm.vm.message.Message`:
- origin: None
- depth: 0
- create_address=None
- code_address=None
- should_transfer_value=True
- is_static=False

**Comment by pipermerriam** [comment]
gas should probably default to the current lock gas limit.

everything else looks good.

**Comment by lrettig** [comment]
Question: how much validation do we need to do on these args? E.g., do we care about gas metering or that the sender account contains at least the value being sent (plus gas)?

**Comment by pipermerriam** [comment]
I think that it's fine to do minimal validation (or near zero validation).  If they send value, we could credit the `to` address, but only deduct from the `sender` if they have appropriate funds.  Or we could just not do the transfer at all, or we could only credit the `to` address....  I'm not sure what is right because I'm not sure all the ways this would be used.  Maybe better to lay out the known use cases and then design with those in mind.

**Comment by lrettig** [comment]
> Maybe better to lay out the known use cases and then design with those in mind.

Agreed. This is a good question. Using ethereumjs's [runCall](https://github.com/ethereumjs/ethereumjs-vm/blob/master/lib/runCall.js) as a model, it looks like that method is used internally (all transactions are passed through it for execution), and it's also called directly in the tests in a bunch of places. There was some additional discussion in https://github.com/ethereumjs/ethereumjs-vm/issues/41 but no conclusions there. I just asked about this in https://gitter.im/ethereum/ethereumjs-lib, let's see what the JS team thinks.

@cwinsloweaton what sort of use cases do you envision for this?

**Comment by pipermerriam** [comment]
@lrettig see https://github.com/ethereum/EIPs/issues/144

**Comment by lrettig** [comment]
Is there a connection between `runCall` and ethereum/EIPs#144?

**Comment by cwinsloweaton** [comment]
@lrettig The use case I had in mind was using it to analyze smart contracts that don't have the source code available. I used the ethereumjs version in a script to confirm that different input data sent to a contract actually output a different result. I'm still pretty new to this stuff so hopefully that makes sense!

**Comment by pipermerriam** [comment]
@lrettig 

> Is there a connection between `runCall` and ethereum/EIPs#144?

Yes (at least in the abstract).

Lets say I have a contract

```
contract Secret {
    uint private secret;

    function setSecret(uint value) public {
        secret = value;
    }
}
```

The variable `secret` isn't actually secret, but rather, just not exposed through any getter function.  I may want to compose a custom piece of bytecode which could be run in the context of a deployed version of this contract which returns the `secret` value.

This same example can then be extrapolated out to more complex use cases like, writing custom code to dig into the internal data of a contract and do *something* that isn't present in the existing contract's bytecode.

Thoughts?

**Comment by lrettig** [comment]
Makes sense. Effectively by setting the `code_address` (in the `evm.vm.message.Message` constructor, or equivalently, `address` in ethereumjs's `runCode` params) you can run arbitrary code in the context of (i.e., with full access to) the state of another contract. Am I understanding this correctly?

This should already be possible with ethereumjs's `runCode`.

**Comment by lrettig** [comment]
@cwinsloweaton thanks, that's a great use case and it's very helpful. The only thing I'd add is that, technically, for this use case, you do `not` need to specify the _code_ as an argument since that could be read from state. Also, it sounds like you're just looking at output, not state changes, am I correct?

@pipermerriam do you have any concerns with allowing arbitrary code here?

**Comment by pipermerriam** [comment]
> @pipermerriam do you have any concerns with allowing arbitrary code here?

No concerns.  I think being able to run arbitrary code in the context of an existing contract is valuable.  (But we don't allow any state changes to be persisted, everything should be reverted after execution).

**Comment by lrettig** [comment]
Closed by #315
